src/main/java/com/direwolf20/buildinggadgets/common/building/placement/OriginWrapper.java

```diff
@@ -0,0 +1,37 @@
+package com.direwolf20.buildinggadgets.common.building.placement;
+
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.util.math.BlockPos;
+
+/**
+ * Wraps an {@link IBlockProvider} such that all access to the provider will be translated by the given amount as the
+ * BlockPos passed into the constructor.
+ */
+public final class OriginWrapper implements IBlockProvider {
+
+    private final IBlockProvider provider;
+    private final BlockPos origin;
+
+    /**
+     * @implNote you cannot create a object and pass itself as a parameter of the constructor, so we are safe from looping
+     */
+    public OriginWrapper(IBlockProvider provider, BlockPos origin) {
+        this.provider = provider;
+        this.origin = origin;
+    }
+
+    @Override
+    public BlockPos getOrigin() {
+        return origin;
+    }
+
+    /**
+     * Redirects the call to the wrapped IBlockProvider.
+     */
+    @Override
+    public IBlockState at(BlockPos pos) {
+        // This should be fine since the parameter is @NoBorrow
+        return provider.at(pos.add(origin));
```

Comment:
this would be fine even without (@)NoBorrow... Why should it make Problems?

Reply:
IPlacementTarget might yield a MutableBlockPos depend on which method you call, and if somehow somebody decide to create an implementation that stores a reference to it, it might create problems。

Reply:
But add will always return an immutable copy.

Reply:
The only case where I could think of this causing Problems is, when someone deliberatly creates a class which doesn't return a copy from add...